pkg/river/parser/internal.go

```diff
@@ -0,0 +1,614 @@
+		} else {
+			p.addErrorf("expected block body, got %s", p.tok)
+		}
+
+		// Give up on this statement and skip to the next identifier.
+		p.advance(token.IDENT)
+		return nil
+	}
+}
+
+// parseBlockName parses the name used for a block.
+//
+//     BlockName = identifier { "." identifier } [ string ]
+func (p *parser) parseBlockName() *blockName {
+	if p.tok != token.IDENT {
+		p.addErrorf("expected identifier, got %s", p.tok)
+		return nil
+	}
+
+	var bn blockName
+
+	bn.Fragments = append(bn.Fragments, p.lit) // Append first identifier
+	bn.Start = p.pos
+	p.next()
+
+	// { "." identifier }
+	for p.tok == token.DOT {
+		p.next() // consume "."
+
+		if p.tok != token.IDENT {
+			p.addErrorf("expected identifier, got %s", p.tok)
+
+			// Continue here to parse as much as possible, even though the block name
+			// will be malformed.
+		}
+
+		bn.Fragments = append(bn.Fragments, p.lit)
+		p.next()
+	}
+
+	// [ string ]
+	if p.tok != token.ASSIGN && p.tok != token.LCURLY {
+		if p.tok == token.STRING && len(p.lit) > 2 {
+			bn.Label = p.lit[1 : len(p.lit)-1] // Strip quotes from label
+			bn.LabelPos = p.pos
+		} else {
+			p.addErrorf("expected block label, got %s", p.tok)
+		}
+		p.next()
+	}
+
+	return &bn
+}
+
+type blockName struct {
+	Fragments []string // Name fragments (i.e., `a.b.c`)
+	Label     string   // Optional user label
+
+	Start    token.Pos
+	LabelPos token.Pos
+}
+
+// ValidAttribute returns true if the blockName can be used as an attribute
+// name.
+func (n blockName) ValidAttribute() bool {
+	return len(n.Fragments) == 1 && n.Label == ""
+}
+
+// ParseExpression parses a single expression.
+//
+//     Expression = BinOpExpr
+func (p *parser) ParseExpression() ast.Expr {
+	return p.parseBinOp(1)
+}
+
+// parseBinOp is the entrypoint for binary expressions. If there is no binary
+// expressions in the current state, a single operand will be returned instead.
+//
+//     BinOpExpr = OrExpr
+//     OrExpr    = AndExpr { "||"   AndExpr }
+//     AndExpr   = CmpExpr { "&&"   CmpExpr }
+//     CmpExpr   = AddExpr { cmp_op AddExpr }
+//     AddExpr   = MulExpr { add_op MulExpr }
+//     MulExpr   = PowExpr { mul_op PowExpr }
+//
+// parseBinOp avoids the need for multiple nonterminal functions by providing
+// context for operator precedence in recursive calls. inPrec specifies the
+// incoming operator precedence. On the first call to parseBinOp, inPrec should
+// be 1.
+//
+// parseBinOp can only handle left-associative operators, so PowExpr is handled
+// by parsePowExpr.
+func (p *parser) parseBinOp(inPrec int) ast.Expr {
+	// The EBNF documented by the function can be generalized into:
+	//
+	//     CurPrecExpr = NextPrecExpr { cur_prec_ops NextPrecExpr }
+	//
+	// The code below implements this specific grammar, continually collecting
+	// everything at the same precedence level into the LHS of the expression
+	// while recursively calling parseBinOp for higher-precedence operations.
+
+	lhs := p.parsePowExpr()
```

Comment:
@rfratto sorry for missing this yesterday, but I think this could have a better name in the future.

It _does_ handle pow expressions that need to shift around the lhs and rhs, but is also the entrypoint for all unary/primary expressions so it had me confused for a while.